Segment processing framework

npawar opened the pull request:
## Description
https://github.com/apache/incubator-pinot/issues/5753
A Segment Processing Framework to convert "m" segments to "n" segments
The phases of the Segment Processor are
1. **Map**
- RecordTransformation (using transform functions)
- Record filtering (using filter functions)
- Partitioning (Column value based, transform function based, table config's partition config based)
2. **Reduce**
- Rollup/Concat records
- Split into parts 
- Sort
3. **Segment generation**

A SegmentProcessorFrameworkCommand is provided to run this on demand. Run using command
`bin/pinot-admin.sh SegmentProcessorFramework -segmentProcessorFrameworkSpec /<path>/spec.json` 
where spec.json is
```
{
  "inputSegmentsDir": "/<base_dir>/segmentsDir",
  "outputSegmentsDir": "/<base_dir>/outputDir/",
  "schemaFile": "/<base_dir>/schema.json",
  "tableConfigFile": "/<base_dir>/table.json",
  "recordTransformerConfig": {
    "transformFunctionsMap": {
      "epochMillis": "round(epochMillis, 86400000)" // round to nearest day
    }
  },
  "recordFilterConfig": {
    "recordFilterType": "FILTER_FUNCTION",
    "filterFunction": "Groovy({epochMillis != \"1597795200000\"}, epochMillis)"
  },
  "partitioningConfig": {
    "partitionerType": "COLUMN_VALUE", // partition on epochMillis
    "columnName": "epochMillis"
  },
  "collectorConfig": {
    "collectorType": "ROLLUP", // rollup clicks by summing
    "aggregatorTypeMap": {
      "clicks": "SUM"
    }
  },
  "segmentConfig": {
    "maxNumRecordsPerSegment": 200_000
  }
}
```

Note:
1. Currently this framework attempts to do no parallelism in the map/reduce/segment creation jobs. Each input file will be processed sequentially in map stage, each part will be executed sequentially in reduce, and each segment will be built one after another. We can change this in the future if the need arises to make this more advanced.
2. The framework makes the assumption that there's enough memory to hold all records of a partition in memory, during rollups in reducer. A limit of 5M records has been set on the Reducer as the number of records to collect before forcing a flush, as a safety measure. In future we could consider using off heap processing, if memory becomes a problem.

 This framework will typically be used by minion tasks, which want to perform some processing on segments
 (eg task which merges segments, tasks which aligns segments per time boundaries etc). The existing Segment merge jobs can be changed to use this framework.

**Pending**
Enhancements like (Added TODOs in code)
- Put null in GenericRecord if nullValueFields contains the field
- Interface out underlying file format (currently avro)
- Dedup
- Using off-heap based implementation for aggregation/sorting in the reduce
- 2 step partitioner 1) Apply custom partitioner 2) Apply table config partitioner. Combine both to get final partition.
- Configs for segment name (like prefix)

snleee commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/framework/SegmentMapper.java:
```diff
@@ -0,0 +1,148 @@
+
+    PinotSegmentRecordReader segmentRecordReader = new PinotSegmentRecordReader(_inputSegment);
+    GenericRow reusableRow = new GenericRow();
+    GenericData.Record reusableRecord = new GenericData.Record(_avroSchema);
```
Can we interface out the underlying data file format? We may need to plug in other data format for performance boost (for example, @Jackie-Jiang used the mmaped file with custom format for star-tree generator and it's also doing the similar work - sort, aggregation)

snleee commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/transformer/TransformFunctionRecordTransformer.java:
```diff
@@ -0,0 +1,49 @@
+
+  private final Map<String, FunctionEvaluator> _functionEvaluatorMap = new HashMap<>();
+
+  public TransformFunctionRecordTransformer(Map<String, String> transformFunctionMap) {
```
What if I want to run the custom transformation function?

The spec doesn't have a way to specify the custom transformer?

```
  "recordTransformerConfig": {
    "transformFunctionsMap": {
      "epochMillis": "round(epochMillis, 86400000)" // round to nearest day
    }
  },

```

npawar replied:
Several options:
1. Use Groovy function to write any logic
2. Add the function with @ScalarFunction annotation and put the jar in plugins dir
3. Introduce new RecordTransformer implementation, introduce RecordTransformerType and change factory logic 

I did not do the option 3 right now, because I think 1 and 2 should suffice most times. If we see the demand for 3, we can add it. 
Will put a TODO

snleee commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/framework/SegmentMapper.java:
```diff
@@ -0,0 +1,148 @@
+      reusableRow = segmentRecordReader.next(reusableRow);
+
+      // Record transformation
+      reusableRow = _recordTransformer.transformRecord(reusableRow);
```
How do we handle `record filtering` that happens in the `transformRecord`?

Or, are we model filtering as a separate step? We should have some convention on record filtering.

npawar replied:
Currently no record filtering happens in transformRecord. If we want to do record filtering in Mapper, I think we can introduce a new phase, then it would become  transform -> filter -> partition -> partitionFilter -> flush

My goal in this phase was to mainly include everything that the current code (of segment merge/rollup) does, and also include everything that the realtime to offline project needs. Filtering wasn't required in either, so skipped it

npawar replied:
I have removed PartitionFilter step and introduced record filtering.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/partitioner/PartitionerFactory.java:
```diff
@@ -0,0 +1,82 @@
+      case NO_OP:
+        partitioner = new NoOpPartitioner();
+        break;
+      case ROW_HASH:
```
I feel ROW_HASH is not really useful. Maybe ROUND_ROBIN is enough?

npawar replied:
The intention of ROW_HASH is to have a partitioner which lets you set number of partitions. The existing SegmentConverter supports only this type of partitioning. Agree that ROW_HASH is not very indicative of the intent.
How about NUM_PARTITIONS or FIXED_NUM_PARTITIONS

Jackie-Jiang replied:
I understand the intention for this partitioner. My suggestion is to not use the hashcode of the record, but simply do round robin on the configured `_numPartitions` which guarantees the records for each partition are balanced. This is also much cheaper comparing to the hashcode based partitioner.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/collector/CollectorConfig.java:
```diff
@@ -0,0 +1,85 @@
+  private final CollectorFactory.CollectorType _collectorType;
+  private final Map<String, ValueAggregatorFactory.ValueAggregatorType> _aggregatorTypeMap;
+
+  private CollectorConfig(CollectorFactory.CollectorType collectorType,
```
It should also include sorted columns (list of columns, where records are sorted on the first column firstly, then second column etc.)

npawar replied:
Added the config and support for sorting in the collection.
Kept the implementation very simple right now. Since we already have all records in memory due to aggregation step, simply called list.sort.
Lmk if you think any optimizations are needed there

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/framework/SegmentReducer.java:
```diff
@@ -0,0 +1,151 @@
+        _collector.collect(next);
+
+        // Exceeded max records allowed to collect. Flush
+        if (_collector.size() == MAX_RECORDS_TO_COLLECT) {
```
Should we use `_numRecordsPerPart` here so that once the collector collects enough records, we flush them?

mcvsubbu replied:
Instead of checking the number of records, can we have a method in the collector that says it is full? The criteria could then be something else depending on the collector.

npawar replied:
`Should we use _numRecordsPerPart here so that once the collector collects enough records, we flush them` - I had initially done it this way. But then I noticed while testing, that if user sets a very low numRecordsPerPart, we will aggregate less and flush very frequently. Having a bigger MAX_RECORDS_TO_COLLECT allows for more aggregation

Jackie-Jiang replied:
I think it is fine if user configures a very low `_numRecordsPerPart` and not get much aggregation. The problem of using the fixed large threshold is that for use cases with lots of columns, we might run out of memory and there is no knob to tune it.
In order to get better aggregation, we can potentially do an external sort on the files before collecting the records. Not requires in the first version.

npawar replied:
sounds good. Changed it to flush when numRecordsPerPart is reached.
@mcvsubbu we can certainly do that if we start seeing need for more advanced controls and tuning on collection size and flushing. For starters, I wanted to keep it simple

mcvsubbu commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/collector/CollectorFactory.java:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  public enum CollectorType {
+    ROLLUP, CONCAT
```
It is useful to add some comments on each of these, explaining what the collector does (or, what it does differently than others)

mcvsubbu commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/collector/RollupCollector.java:
```diff
@@ -0,0 +1,138 @@
+  private final MetricFieldSpec[] _metricFieldSpecs;
+
+  public RollupCollector(CollectorConfig collectorConfig, Schema schema) {
+    _keySize = schema.getColumnNames().size() - schema.getMetricNames().size();
```
So, we include virtual columns here?

npawar replied:
nope, changed it.

mcvsubbu commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/collector/RollupCollector.java:
```diff
@@ -0,0 +1,138 @@
+   * Note that the dimensions can have multi-value columns, and hence the equals and hashCode need deep array operations
+   */
+  private static class Record {
+    private final Object[] _values;
```
Are these the key parts? If so, can we name the member as `_keyParts`? `_values` gets confusing.

mcvsubbu commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/partitioner/PartitionerFactory.java:
```diff
@@ -0,0 +1,82 @@
+            "Must provide transformFunction for TRANSFORM_FUNCTION partitioner");
+        partitioner = new TransformFunctionPartitioner(config.getTransformFunction());
+        break;
+      case TABLE_PARTITION_CONFIG:
```
Since these are all derived segments (merging m segments into n) should we not be using the table's partitioner all the time?

npawar replied:
Yes. Seunghyun pointed out similar thing. Discussion below.
I'll be changing the partitioner to a 2 step partitioner in the next PR (have put TODO in code and descritpion)

mcvsubbu commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/utils/SegmentProcessorUtils.java:
```diff
@@ -0,0 +1,113 @@
+  /**
+   * Convert a GenericRow to an avro GenericRecord
+   */
+  public static GenericData.Record convertGenericRowToAvroRecord(GenericRow genericRow,
```
These should go into avro plugins? Why introduce dependency on avro here?

npawar replied:
avro plugins is more for the extractors and decoders right. This is a util method. And pinot-core cannot be made to depend on pinot-avro for this

npawar commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/partitioner/PartitionFilter.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Used for filtering partitions in the mapper
+ */
+public interface PartitionFilter {
```
true. Remove partition filtering. Added Record filtering. All the corresponding changes in FunctionEvaluator have been reverted.

npawar commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/collector/Collector.java:
```diff
@@ -0,0 +1,50 @@
+  Iterator<GenericRow> iterator();
+
+  /**
+   * The size of the collection
```
clarified in javadoc. it is num records

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/framework/SegmentMapper.java:
```diff
@@ -0,0 +1,135 @@
+      String partition = _partitioner.getPartition(reusableRow);
+
+      // Create writer for the partition, if not exists
+      if (!_partitionToDataFileWriterMap.containsKey(partition)) {
```
Use `_partitionToDataFileWriterMap.get(partition)` and check if the value is null to save one map lookup

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/framework/SegmentProcessorConfig.java:
```diff
@@ -0,0 +1,181 @@
+    public SegmentProcessorConfig build() {
+      Preconditions.checkNotNull(tableConfig, "Must provide table config in SegmentProcessorConfig");
+      Preconditions.checkNotNull(schema, "Must provide schema in SegmentProcessorConfig");
+      if (recordTransformerConfig == null) {
```
Suggest leaving them as `null` if not set

npawar replied:
i just wanted to avoid checking null before using them in the mapper and reducer

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/utils/SegmentProcessorUtils.java:
```diff
@@ -0,0 +1,113 @@
+   */
+  public static GenericData.Record convertGenericRowToAvroRecord(GenericRow genericRow,
+      GenericData.Record reusableRecord) {
+    for (String field : genericRow.getFieldToValueMap().keySet()) {
```
Put `null` values from `GenericRow.getNullValueFields()`?

npawar replied:
I didn't follow. Aren't all nullValueFields already expected to be in fieldToValueMap ?

Jackie-Jiang replied:
The value in `fieldToValueMap` should be the default value if the original value is `null`. IMO we should put `null` instead of default value in the `GenericData.Record` so that the `null` values can be populated to the new segment

npawar replied:
I see, got it. In the interest of getting this long standing and big PR merged, I will think about this and take it up immediately in a following PR.

Jackie-Jiang reviewed (approved):
Good job splitting the work into multiple modules and make it very easy to extend. Mostly minor comments.
One high level comment: since we always use Avro as the intermediate format, should we directly work on `GenericRecord` instead of converting back and forth between `GenericRow` and `GenericRecord`?
Also, we might want to support more input formats other than Pinot segments. We can do it as the next step.

snleee commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/collector/Collector.java:
```diff
@@ -0,0 +1,55 @@
+   * Collects the given GenericRow and stores it
+   * @param genericRow the generic row to add to the collection
+   */
+  void collect(GenericRow genericRow);
```
I personally prefer GenericRow to allow the extension of using file formats other than Avro.

npawar replied:
I also felt that GenericRow is better. I was aiming for consistency between Mapper and Reducer. Seunghyun's point also makes sense.

snleee commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/partitioner/NoOpPartitioner.java:
```diff
@@ -0,0 +1,32 @@
+public class NoOpPartitioner implements Partitioner {
+  @Override
+  public String getPartition(GenericRow genericRow) {
+    return "0";
```
No-op partitioner means that we always create a single output file?

npawar replied:
That is correct. A single output from each Mapper. The reducer will break that data as per `maxRecordsPerSegment`

snleee commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/partitioner/ColumnValuePartitioner.java:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  @Override
+  public String getPartition(GenericRow genericRow) {
```
Is this intended for supporting time alignment?

What if the time column granularity is in seconds/hours while push frequency is DAY?

In that case, we may need to use `TransformationPartitioner`?

npawar replied:
Not particularly written for supporting time alignment, but can be used for that if values are to be used as is for partitioning.

If time column is seconds/hours, but we want to align by day, yes we can use TransformFunctionPartitioner.

snleee commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/partitioner/TableConfigPartitioner.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Partitioner which computes partition values based on the ColumnPartitionConfig from the table config
+ */
+public class TableConfigPartitioner implements Partitioner {
```
What if I need to align data on time while the table is custom partitioned? (just trying to brainstorm how we will extend the current partitioner to support this)

Then, we can probably add the new partitioner that combines the value from `TableConfigPartitioner` and `TransformationPartitioner`?


e.g.  partition on memberId using murmur, need to enable segment merge so the data needs to be time aligned.

1. Use table config partitioner to get the partition id based on murmur on memberId -> Let's day `2`
2. Use time align partitioner -> Let's say `2020/12/12`

Combine 1&2 -> `2020/12/12-2` <- example of partitionId

We can do something like the above?

npawar replied:
this is a good point. I did not think of this. Yes we can add new partitioner in that case, which combines the partitioners as needed.

But this might be a common case, so I'm thinking if we should have that in the design itself. How about we always do partitioning in 2 steps inside the mapper. 
1) Apply any partitioning from Segment Processor Config 
2) Apply any partitioning from Table Config. 
So in your example, first TransformFunctionPartitioner gets applied and generated date partition. Then, we check if table config partitioner exists, and if yes, further apply the PartitionFunction. Then concat the 2 partitions like you suggested.

Another option is we make PartitionerConfig a list, and apply all partitioners one by one, and concat all to get final partition.

snleee commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/framework/SegmentProcessorConfig.java:
```diff
@@ -0,0 +1,181 @@
+/**
+ * Config for configuring the phases of {@link SegmentProcessorFramework}
+ */
+public class SegmentProcessorConfig {
```
One requirement for `SegmentMergeRollup` is to be able to put the custom name for the segment name (or at least need to put the prefix and the sequenced `merged_XXX_0...M`

Where do you think it's the best place to configure those?

npawar replied:
These config can go into the `SegmentConfig` class.
I will add seqId to the SegmentGenerationConfig in the driver. seqId will also help with the other problem you caught (same segment name if start/end is same). 

Config for prefix can be added in a future change.

snleee commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/framework/SegmentProcessorFramework.java:
```diff
@@ -0,0 +1,194 @@
+    // Segment generation phase.
+    LOGGER.info("Beginning segment generation phase. Processing files: {}", Arrays.toString(_reducerOutputDir.list()));
+    // Reducer output directory will have 1 or more avro files
+    for (File resultFile : reducerOutputFiles) {
```
Did you check the output segment names when the output is more than 1 files?

It's possible that the final segments may have the same segment name. (e.g. `<tablename>_<start>_<end>`)

npawar replied:
Good catch. Will add seqId. Also working on adding end-to-end tests for the framework, and will include this case

snleee reviewed (commented):
Overall, I like that all the core components are interfaced out and easy to extend. I have put some comments. Some of them are questions or points that I would like to discuss.

npawar commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/collector/ValueAggregator.java:
```diff
@@ -16,7 +16,7 @@
  * under the License.
  */
-package org.apache.pinot.core.minion.rollup.aggregate;
+package org.apache.pinot.core.segment.processing.collector;
```
Changed.

npawar commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/collector/CollectorConfig.java:
```diff
@@ -0,0 +1,103 @@
+/**
+ * Config for Collector
+ */
+@JsonDeserialize(builder = CollectorConfig.Builder.class)
```
Agreed about JsonCreator. Changed it.

npawar commented on pinot-core/src/main/java/org/apache/pinot/core/segment/processing/collector/RollupCollector.java:
```diff
@@ -0,0 +1,159 @@
+   * If a row already exists in the collection (based on dimension + time columns), rollup the metric values, else add the row
+   */
+  @Override
+  public void collect(GenericRow genericRow) {
```
Yes, we keep the entire data of a partition on heap.
Yes, we can certainly add off-heap or file-based in future. For now, we have a knob `maxRecordsPerSegment` that can help control number of records collected in memory.

In order to sort the entire data on all dimensions+time, we'll still have to get all the data into memory right? And then we'll have to rewrite the files onto disk. The input is raw avro file. We do not have any dictionaries or docIds.

npawar commented:
> Overall, I like that all the core components are interfaced out and easy to extend. I have put some comments. Some of them are questions or points that I would like to discuss.

Addressed the comments. Added TODOs in code and description for those that will be handled in future PRs

snleee reviewed (approved):
LGTM. Thank you for addressing all the comments!
